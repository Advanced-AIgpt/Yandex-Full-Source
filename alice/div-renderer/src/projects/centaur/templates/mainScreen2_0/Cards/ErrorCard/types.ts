import { IAbstractCardProps } from '../types';

export interface IErrorCardProps extends IAbstractCardProps {
    type: 'error';
}
