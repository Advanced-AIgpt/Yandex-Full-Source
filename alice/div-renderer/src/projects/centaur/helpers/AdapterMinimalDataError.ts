export class AdapterMinimalDataError extends Error {
    constructor(missedData: string[], template: string) {
        super(`Missed data (${missedData.join(', ')}) in template ${template}`);
        this.name = 'AdapterMinimalDataError';
    }
}
