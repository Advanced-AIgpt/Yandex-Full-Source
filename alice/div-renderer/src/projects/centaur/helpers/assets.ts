export function getS3Asset(path: string) {
    return `https://dialogs.s3.yandex.net/smart_displays/${path}`;
}

export function getStaticS3Asset(path: string) {
    return `https://yastatic.net/s3/dialogs/smart_displays/${path}`;
}
