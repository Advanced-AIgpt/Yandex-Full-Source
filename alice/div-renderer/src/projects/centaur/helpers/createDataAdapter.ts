import Ajv, { AnySchema, JSONSchemaType } from 'ajv';
import { DeepNonNullable, DeepRequired } from 'ts-essentials';
import { IRequestState } from '../../../common/types/common';
import { ExpFlags } from '../expFlags';

export const ajv = new Ajv();

type DataAdapter<D, R> = (data: D, requestState: IRequestState) => R;

export const createDataAdapter = <D, R>(schema: AnySchema, dataAdapter: DataAdapter<D, R>) => {
    const validate = ajv.compile(schema);

    const resultDataAdapter: DataAdapter<D, R> = (data, requestState) => {
        const isDataValid = validate(data);

        if (isDataValid === false) {
            // TODO: CENTAUR-1243: log errors

            if (requestState.hasExperiment(ExpFlags.errorOnAssert)) {
                const [firstError] = validate.errors ?? [];

                throw new Error(`DataValidationError: "${firstError.instancePath}" ${firstError.message}`);
            }
        }

        return dataAdapter(data, requestState);
    };

    return resultDataAdapter;
};

export type MakeClear<T> = DeepRequired<DeepNonNullable<T>>;
export type MakeTestSchemaType<T> = JSONSchemaType<MakeClear<T>>;
