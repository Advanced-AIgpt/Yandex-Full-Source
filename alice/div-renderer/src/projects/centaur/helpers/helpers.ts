/* eslint-disable camelcase */
import { sortedIndexBy } from 'lodash';
import { format, isSameDay } from 'date-fns';
import { ru } from 'date-fns/locale';
import { Div, IDivData, TextBlock } from 'divcard2';
import { IRequestState } from '../../../common/types/common';
import { getTemplates, templatesRegistry } from '../../../registries/common';
import { centaurTemplatesClass } from '../index';
import { Palette } from './palette';

export { default as pluralize } from '@yandex-int/i18n/lib/plural/ru';
type TextProps = Pick<ConstructorParameters<typeof TextBlock>[0], 'text_color'|'font_size'|'font_weight'|'line_height'>;

export const textBreakpoint =
    (...breakpoints: [number, TextProps][]) =>
        (text: string) => {
            const idx = Math.min(sortedIndexBy(breakpoints, [text.length, {}], 0), breakpoints.length - 1);
            return { text, ...breakpoints[idx][1] };
        };

export const formatDate = (d: Date | string | number, f: string) => format(new Date(d), f, { locale: ru });

export const textCommand = (c: string) => `dialog://text_command?query=${encodeURI(c)}`;

export const updatedAt = (d: Date) => {
    const now = new Date();
    return `Обновлен ${isSameDay(now, d) ? 'сегодня' : formatDate(d, 'dd.MM')}`;
};

export const insertBetween = <I extends unknown, S extends unknown>(itemList: I[], separator: S) => {
    return itemList.reduce<(I | S)[]>((result, item, index) => {
        if (index !== 0) {
            result.push(separator);
        }

        result.push(item);

        return result;
    }, []);
};

export interface ITemplateCard {
    templates: {
        [name: string]: Div;
    };
    palette?: Palette;
    card: IDivData;
}

// Дива верхнего уровня. Замена для new `TemplateCard()`
export const TopLevelCard = (card: Omit<IDivData, 'variable_triggers' | 'variables'>, requestState: IRequestState, palette?: Palette) => {
    const templates = Object.assign(
        getTemplates(requestState.localTemplates, templatesRegistry),
        centaurTemplatesClass.getUsedTemplates(requestState),
        requestState.res.localTemplates.getAll(),
    );

    return {
        templates,
        palette,
        card: {
            ...card,
            variable_triggers: requestState.variableTriggers.getAll(),
            variables: [...requestState.variables],
        } as IDivData,
    };
};
