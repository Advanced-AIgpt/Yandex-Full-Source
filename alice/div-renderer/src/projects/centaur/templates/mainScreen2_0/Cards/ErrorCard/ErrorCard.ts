import { Div } from 'divcard2';
import BasicErrorCard from '../BasicErrorCard';

export default function ErrorCard({ colIndex, rowIndex }: {colIndex: number | undefined; rowIndex: number | undefined }): Div {
    return BasicErrorCard({
        colIndex,
        rowIndex,
        title: 'Ошибка',
        description: 'Что-то пошло не так. Проверьте интернет или обновите экран',
    });
}
