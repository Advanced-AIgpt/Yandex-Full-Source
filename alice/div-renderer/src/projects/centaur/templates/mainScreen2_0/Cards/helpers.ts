export function getCardMainScreenId({ rowIndex, colIndex }: { rowIndex?: number; colIndex?: number }) {
    return `main_screen_card_${rowIndex || 0}_${colIndex || 0}`;
}
