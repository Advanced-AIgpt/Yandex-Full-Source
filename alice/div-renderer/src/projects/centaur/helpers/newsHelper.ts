import { getStaticS3Asset } from './assets';

const TOPICS = new Set([
    'auto',
    'business',
    'computers',
    'culture',
    'incident',
    'index',
    'koronavirus',
    'personal',
    'politics',
    'science',
    'society',
    'sport',
    'world',
]);

const ALIASES = {
    nplus1: 'science',
} as { [name: string]: string };

export function getImageByTopic(topic: string | null | undefined): string {
    let resultImage = 'news/index.png';

    if (topic) {
        if (Object.prototype.hasOwnProperty.call(ALIASES, topic)) {
            topic = ALIASES[topic];
        }

        if (TOPICS.has(topic)) {
            resultImage = `news/${topic}.png`;
        }
    }
    return getStaticS3Asset(resultImage);
}
