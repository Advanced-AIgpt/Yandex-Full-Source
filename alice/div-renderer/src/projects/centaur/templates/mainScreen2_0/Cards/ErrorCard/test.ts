import ErrorCard from './ErrorCard';
import { AnonymizeDataForSnapshot } from '../../../../../../common/helpers/dev';

describe('Main screen empty card', () => {
    it('should match snapshot', () => {
        expect(AnonymizeDataForSnapshot(ErrorCard({ colIndex: 0, rowIndex: 0 }))).toMatchSnapshot();
    });
});
