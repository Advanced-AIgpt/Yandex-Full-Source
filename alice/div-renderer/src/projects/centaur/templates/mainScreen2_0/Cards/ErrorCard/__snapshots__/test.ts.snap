// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main screen empty card should match snapshot 1`] = `
Object {
  "background": Array [
    Object {
      "color": "#1aeaf6ec",
      "type": "solid",
    },
  ],
  "border": Object {
    "corner_radius": 28,
  },
  "height": Object {
    "type": "match_parent",
    "weight": 1,
  },
  "id": "main_screen_card_0_0",
  "items": Array [
    Object {
      "items": Array [
        Object {
          "font_size": 32,
          "font_weight": "medium",
          "line_height": 40,
          "text": "Ошибка",
          "text_color": "#80ffffff",
          "type": "text",
        },
      ],
      "orientation": "horizontal",
      "type": "container",
    },
    Object {
      "delimiter_style": Object {
        "color": "#0fff",
      },
      "height": Object {
        "type": "match_parent",
      },
      "type": "separator",
    },
    Object {
      "font_size": 28,
      "font_weight": "medium",
      "line_height": 36,
      "text": "Что-то пошло не так. Проверьте интернет или обновите экран",
      "text_color": "#80ffffff",
      "type": "text",
    },
  ],
  "paddings": Object {
    "bottom": 24,
    "left": 24,
    "right": 24,
    "top": 24,
  },
  "type": "container",
  "width": Object {
    "type": "match_parent",
    "weight": 1,
  },
}
`;
